bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java

```diff
@@ -115,121 +120,150 @@
+                newProductUrlFormat = new ProductPageUrlFormatAdapter(productPageUrlFormat);
+            } else {
+                newProductUrlFormat = DEFAULT_PRODUCT_URL_FORMATS
+                    .getOrDefault(conf.productPageUrlFormat(), ProductPageWithUrlKey.INSTANCE);
+            }
         }
-        if (categoryPageUrlFormat == null) {
-            categoryPageUrlFormat = DEFAULT_CATEGORY_URL_FORMATS
-                .getOrDefault(conf.categoryPageUrlFormat(), CategoryPageWithUrlPath.INSTANCE);
+        if (newCategoryUrlFormat == null) {
+            if (categoryPageUrlFormat != null) {
+                newCategoryUrlFormat = new CategoryPageUrlFormatAdapter(categoryPageUrlFormat);
+            } else {
+                newCategoryUrlFormat = DEFAULT_CATEGORY_URL_FORMATS
+                    .getOrDefault(conf.categoryPageUrlFormat(), CategoryPageWithUrlPath.INSTANCE);
+            }
         }
     }
 
     @Deactivate
     protected void deactivate() {
         productPageUrlFormat = null;
+        newProductUrlFormat = null;
         categoryPageUrlFormat = null;
+        newCategoryUrlFormat = null;
     }
 
     @Override
     public String toProductUrl(SlingHttpServletRequest request, Page page, Map<String, String> params) {
-        return toUrl(request, page, params, productPageUrlFormat);
+        return toProductUrl(request, page, new ProductUrlFormat.Params(params));
     }
 
     @Override
     public String toProductUrl(SlingHttpServletRequest request, Page page, String productIdentifier) {
-        ParamsBuilder params = new ParamsBuilder();
+        ProductUrlFormat.Params params = new ProductUrlFormat.Params();
         if (StringUtils.isNotBlank(productIdentifier)) {
-            params.sku(productIdentifier);
-
-            MagentoGraphqlClient magentoGraphqlClient = request.adaptTo(MagentoGraphqlClient.class);
-
-            // for formats that require url_path or url_key we have to lookup them up
-            Set<String> formatParameters = productPageUrlFormat.getParameterNames();
-            if (magentoGraphqlClient != null &&
-                (formatParameters.contains(URL_KEY_PARAM) || formatParameters.contains(URL_PATH_PARAM))) {
-                ProductUrlParameterRetriever retriever = new ProductUrlParameterRetriever(magentoGraphqlClient);
-                retriever.setIdentifier(productIdentifier);
-                ProductInterface product = retriever.fetchProduct();
-                if (product != null) {
-                    params
-                        .urlKey(product.getUrlKey())
-                        .urlPath(product.getUrlPath());
-                } else {
-                    LOGGER.debug("Could not generate product page URL for {}.", productIdentifier);
+            params.setSku(productIdentifier);
+            // assume that any other format then the ProductPageWithSku requires more parameters
+            if (!(newProductUrlFormat instanceof ProductPageWithSku)) {
```

Comment:
When we last refactored this we introduced the `UrlFormat#getParameterNames()` to allow `UrlFormats` to return which information they need to format a url. This allowed us to bypass the query here when only the sku is needed. However, this is also the only use case I could imagine. Any other format would require at least one more parameter. Due to that I removed the `getParameterNames()` again and replaced it with this `instanceof` check.